test/e2e/apimachinery/crd_validation_rules.go

```diff
@@ -0,0 +1,158 @@
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"context"
+	"strings"
+
+	"github.com/onsi/ginkgo"
+
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apiextensions-apiserver/test/integration/fixtures"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = SIGDescribe("CustomResourceValidationRules [Privileged:ClusterAdmin][Alpha][Feature:CustomResourceValidationExpressions]", func() {
+	f := framework.NewDefaultFramework("crd-validation-expressions")
+
+	var apiExtensionClient *clientset.Clientset
+	ginkgo.BeforeEach(func() {
+		var err error
+		apiExtensionClient, err = clientset.NewForConfig(f.ClientConfig())
+		framework.ExpectNoError(err, "initializing apiExtensionClient")
+	})
+
+	unmarshallSchema := func(schemaJson []byte) *v1.JSONSchemaProps {
+		var c v1.JSONSchemaProps
+		err := json.Unmarshal(schemaJson, &c)
+		framework.ExpectNoError(err, "unmarshalling OpenAPIv3 schema")
+		return &c
+	}
+	var schemaWithValidationExpression = unmarshallSchema([]byte(`{
+	   "type":"object",
+	   "properties":{
+		  "spec":{
+			 "type":"object",
+			 "x-kubernetes-validations":[
+		       { "rule":"self.x + self.y > 0" }
+	         ],
+			 "properties":{
+				"x":{ "type":"integer" },
+				"y":{ "type":"integer" }
+			 }
+		  },
+		  "status":{
+			 "type":"object",
+			 "x-kubernetes-validations":[
+				{ "rule":"self.health == 'ok' || self.health == 'unhealthy'" }
+			 ],
+			 "properties":{
+				"health":{ "type":"string" }
+			 }
+		  }
+	   }
+	}`))
+	ginkgo.It("MUST NOT fail validation for create and update of a custom resource that satisfies the x-kubernetes-validator rules", func() {
+		ginkgo.By("Creating a custom resource definition with validation rules")
+		crd := fixtures.NewRandomNameV1CustomResourceDefinitionWithSchema(v1.NamespaceScoped, schemaWithValidationExpression)
+		crd, err := fixtures.CreateNewV1CustomResourceDefinition(crd, apiExtensionClient, f.DynamicClient)
+		framework.ExpectNoError(err, "creating CustomResourceDefinition")
+
+		defer func() {
+			err = fixtures.DeleteV1CustomResourceDefinition(crd, apiExtensionClient)
+			framework.ExpectNoError(err, "deleting CustomResourceDefinition")
+		}()
+
+		ginkgo.By("Creating a custom resource with values that are allowed by the validation rules set on the custom resource definition")
+		gvrs := fixtures.GetGroupVersionResourcesOfCustomResource(crd)
+		if len(gvrs) == 0 {
+			ginkgo.Fail("Expected at least one version in custom resource definition")
+		}
+		gvr := gvrs[0]
+		crClient := f.DynamicClient.Resource(gvr)
+		name1 := names.SimpleNameGenerator.GenerateName("cr-1")
+		cr, err := crClient.Create(context.TODO(), &unstructured.Unstructured{Object: map[string]interface{}{
```

Comment:
do we need the namespace somewhere?
it seems the rest built that fails in the test doesn't have it , and the crd seems to be namespace scoped
> 	crd := NewRandomNameV1CustomResourceDefinition(apiextensionsv1.NamespaceScoped)


```
crClient.Namespace().Create)
```
or something like that?

Reply:
Good catch, that's what failed. I've fixed the namespacing.  I think this test should be cleaning up all created resources properly now.